### DISCUSSION TITLE
Feature Request: Implement stable sorting in DuckDB

### DISCUSSION THREAD

**Discussion by javafanboy** [duckdb/duckdb#14941] [Ideas]
In analytical processing in particular it is not uncommon with dataset lackin a useful unique key (or even useful unique composite key) and in order to optimize performance it is often important to maintain the native order of data (for instance to enable efficient delta encoding or enable use of algorithms that require a sorted dataset like binary search etc). 

Today DuckDB seemingly maintain native ordering of data both when importing, querying (without sorting) and exporting data but as soon as "order by" is involved the result is no longer "stable" in the sense that rows with identical keys can be arbitrary re-ordered (just tested this with a simple dataset of log lines where the timestamp is non-unique).
Also I have not found it clearly documented that the sorting behavior is "unstable" - as this is an important property of any sorting method it would be good if this was clearly spelled out in documentation of "order by" etc.

As there are stable and highly performant/paralization friendly sorting algorithms available I think it would be great to use one of these or, assuming the current unstable is significantly faster, introduce an option/mode to select stable/unstable sorting.

**Comment by Maxxen** [comment]
SQL in general does not guarantee any implicit ordering of tables or result sets, outside the use of ORDER BY.

Is it not possible to get the desired "stable" behavior by simply adding another column to the ORDER BY clause? DuckDBs sql dialect provides the shorthand "ORDER BY ALL" that might be useful if you always want a consistent ordering.

**Reply by javafanboy** [reply]
I have run into the need for stability in cases where you deal with large datasets you have no control over (produced by third party tools or by hard to change data pipelines) where the only interesting key is often time and the time stamps more or less frequently can be the same on multiple rows as I for instance mentioned in [this](https://github.com/duckdb/duckdb/discussions/14915) discussion.
In these cases it is generally not an option to add more sort columns in addition to the native one of the table (or its reverse that also should be fairly efficient if I understand how DuckDB works) as this internally may "re-sort" the whole gigantic table each time the query is performed (the cost getting worse and worse the more columns that are included). Creating a new table that adds say a rownum column may be possible sometimes but in some case it would consume too much time, disk space etc...
In my view stable sorting is very desirable property in a database as it results in one thing less to worry about when writing queries - things basically work more like you would "naively expect".
At least having selectable stability (sorting algorithm) as an option would imho be valuable and something I would frequently use even if it had a small performance penalty....

**Reply by arjenpdevries** [reply]
Your feature request is actually to implement something different from a relational database engine. You would need an ordered table type; something that may be supported in an array database, but is not supported in the relational model.

(Also, when you select records from a table, there are no guarantees about the order of the records returned; which is actually why select can be implemented at the physical level with, eg, a hash table, to be more efficient. Ie, the result of your stable sort, should you get it, would immediately lose its usefulness as soon as you apply another operator to get data out.)

BTW, more support for ordering support in databases has definitely been a research topic for many years, see eg the 1993 paper by David Maier: https://dl.acm.org/doi/10.1145/153850.153851

**Comment by soerenwolfers** [comment]
To achieve stable sorting, you can add the `rowid` pseudocolumn to your ordering clause.

**Comment by c-herrewijn** [comment]
Note that in some cases order preservation is guaranteed, see: https://duckdb.org/docs/sql/dialect/order_preservation.html
Maybe it is possible to work within these limits.

**Reply by javafanboy** [reply]
This is REALLY helpful had not seen that part of the documentation - just wish it applied for ORDER BY as well (or could be selected to apply as of my proposal)....

**Reply by soerenwolfers** [reply]
what's the problem with adding the rowid column to your sort order?

**Comment by javafanboy** [comment]
Imagine you are not querying a normal native table but rather a bunch of
huge Parquet files in S3 for instance that you can't modify or download in
its entirety then the adding a new column with row numbers is not an option
(unless you are fine with adding it to a view but this results in the cost
of calculating "row numbers" for every query resulting in horrible
performance)...

For a "normal" database operating only on tables (where the proposed
solutions reasonably easily can be implemented) I suppose stable sorting is
not that critical.

Anyhow it seems like nobody but me is doing stuff where this proposal would
make things easier so probably not worth implementing as there are more
requested improvements I suppose....

**Reply by soerenwolfers** [reply]
For parquet files you can add a row number almost for free via `file_row_number`.

In general though, how can you sort something that you can't download in its entirety?

And do you have evidence that adding `rowid` or `row_number() OVER ()` results in horrible performance? You'd only ever notice this on queries where you do use these columns, and I imagine a sort is orders more expensive than this anyway.
